File: src/main/java/dev/boooiil/historia/core/items/component/WeaponComponent.java
package dev.boooiil.historia.core.items.component;

import dev.boooiil.historia.core.items.ItemComponent;
import dev.boooiil.historia.core.items.data.WeaponData;
import dev.boooiil.historia.core.util.JSONUtils;
import dev.boooiil.historia.core.util.NumberUtils;
import org.bukkit.configuration.ConfigurationSection;
import org.jspecify.annotations.NullMarked;

import java.util.List;

@NullMarked
public record WeaponComponent(List<Float> sweepingRange) implements ItemComponent {

    public static WeaponComponent fromConfig(ConfigurationSection section) {
        return new WeaponComponent(section.getFloatList("sweeping"));
    }

    @Override
    public WeaponData data() {
        float sweeping = NumberUtils
                .roundFloat(NumberUtils.random(this.sweepingRange().get(0), this.sweepingRange().get(1)), 2);

        return new WeaponData(sweeping);
    }

    @Override
    public WeaponData data(float qualityModifier) {
        return data();
    }

    @Override
    public String getKey() {
        return "weapon";
    }

    @Override
    public String toString() {

        String sb = "WeaponComponent" +
                toJSON();

        return sb;
    }

    @Override
    public String toJSON() {

        String sb = "{" +
                JSONUtils.fromList("sweepRange", sweepingRange) +
                "}";

        return sb;
    }

}

File: src/main/java/dev/boooiil/historia/core/items/component/WeaponComponent.kt
package dev.boooiil.historia.core.items.component

import dev.boooiil.historia.core.items.ItemComponent
import dev.boooiil.historia.core.items.data.WeaponData
import dev.boooiil.historia.core.util.JSONUtils
import dev.boooiil.historia.core.util.NumberUtils
import org.bukkit.configuration.ConfigurationSection

class WeaponComponent(
    val sweepingRange: MutableList<Float>
) : ItemComponent {

    override val key = "weapon"

    override fun data(): WeaponData {
        val sweeping = NumberUtils.roundFloat(NumberUtils.random(this.sweepingRange[0], this.sweepingRange[1]), 2)

        return WeaponData(sweeping)
    }

    override fun toString(): String {
        val sb = "WeaponComponent" +
                toJSON()

        return sb
    }

    override fun toJSON(): String {
        val sb = "{" +
                JSONUtils.fromList("sweepRange", sweepingRange) +
                "}"

        return sb
    }

    companion object {
        fun fromConfig(section: ConfigurationSection): WeaponComponent {
            return WeaponComponent(section.getFloatList("sweeping"))
        }
    }
}
